File: apps/web/src/lib/api/orgs.ts
import { apiFetch } from "./core"

export interface ApiOrg {
  id: string
  name: string
  slug: string
  created_at: string
  updated_at: string
}

export interface ApiOrgMember {
  id: string
  user_id: string
  role: "owner" | "admin" | "member"
  user_name: string
  user_email: string
}

export interface ApiOrgInvitation {
  id: string
  org_id: string
  email: string
  role: "admin" | "member"
  expires_at: string
  token?: string
}

export interface ApiInvitationInfo {
  email: string
  org_name: string
  role: string
}

export const orgs = {
  list: (token: string) =>
    apiFetch<ApiOrg[]>("/api/v1/orgs", {}, token),

  get: (orgId: string, token: string) =>
    apiFetch<ApiOrg>(`/api/v1/orgs/${orgId}`, {}, token),

  update: (orgId: string, name: string, token: string) =>
    apiFetch<ApiOrg>(`/api/v1/orgs/${orgId}`, { method: "PATCH", body: JSON.stringify({ name }) }, token),

  listMembers: (orgId: string, token: string) =>
    apiFetch<ApiOrgMember[]>(`/api/v1/orgs/${orgId}/members`, {}, token),

  addMember: (orgId: string, email: string, role: "admin" | "member", token: string) =>
    apiFetch<ApiOrgMember>(`/api/v1/orgs/${orgId}/members`, { method: "POST", body: JSON.stringify({ email, role }) }, token),

  updateMember: (orgId: string, userId: string, role: "admin" | "member", token: string) =>
    apiFetch<void>(`/api/v1/orgs/${orgId}/members/${userId}`, { method: "PATCH", body: JSON.stringify({ role }) }, token),

  removeMember: (orgId: string, userId: string, token: string) =>
    apiFetch<void>(`/api/v1/orgs/${orgId}/members/${userId}`, { method: "DELETE" }, token),

  createInvitation: (orgId: string, email: string, role: "admin" | "member", token: string) =>
    apiFetch<ApiOrgInvitation>(`/api/v1/orgs/${orgId}/invitations`, { method: "POST", body: JSON.stringify({ email, role }) }, token),

  listInvitations: (orgId: string, token: string) =>
    apiFetch<ApiOrgInvitation[]>(`/api/v1/orgs/${orgId}/invitations`, {}, token),

  getInvitationByToken: (inviteToken: string) =>
    apiFetch<ApiInvitationInfo>(`/api/v1/invitations/${inviteToken}`),

  acceptInvitation: (inviteToken: string, username: string, password: string) =>
    apiFetch<{ message: string }>(`/api/v1/invitations/${inviteToken}/accept`, {
      method: "POST",
      body: JSON.stringify({ username, password }),
    }),
}

File: apps/cli/internal/client/orgs.go
package client

type OrgMember struct {
	ID        string `json:"id"`
	UserID    string `json:"user_id"`
	Role      string `json:"role"`
	UserName  string `json:"user_name"`
	UserEmail string `json:"user_email"`
}

type Invitation struct {
	ID        string `json:"id"`
	OrgID     string `json:"org_id"`
	Email     string `json:"email"`
	Role      string `json:"role"`
	ExpiresAt string `json:"expires_at"`
	Token     string `json:"token,omitempty"`
}

func (c *Client) ListMembers(orgID string) ([]OrgMember, error) {
	resp, err := c.do("GET", "/api/v1/orgs/"+orgID+"/members", nil)
	if err != nil {
		return nil, err
	}
	return decode[[]OrgMember](resp)
}

func (c *Client) AddMember(orgID, email, role string) (*OrgMember, error) {
	resp, err := c.do("POST", "/api/v1/orgs/"+orgID+"/members", map[string]string{"email": email, "role": role})
	if err != nil {
		return nil, err
	}
	return decodePtr[OrgMember](resp)
}

func (c *Client) UpdateMember(orgID, userID, role string) error {
	return c.doNoContent("PATCH", "/api/v1/orgs/"+orgID+"/members/"+userID)
}

func (c *Client) UpdateMemberRole(orgID, userID, role string) error {
	resp, err := c.do("PATCH", "/api/v1/orgs/"+orgID+"/members/"+userID, map[string]string{"role": role})
	if err != nil {
		return err
	}
	resp.Body.Close()
	return nil
}

func (c *Client) RemoveMember(orgID, userID string) error {
	return c.doNoContent("DELETE", "/api/v1/orgs/"+orgID+"/members/"+userID)
}

func (c *Client) CreateInvitation(orgID, email, role string) (*Invitation, error) {
	resp, err := c.do("POST", "/api/v1/orgs/"+orgID+"/invitations", map[string]string{"email": email, "role": role})
	if err != nil {
		return nil, err
	}
	return decodePtr[Invitation](resp)
}

func (c *Client) ListInvitations(orgID string) ([]Invitation, error) {
	resp, err := c.do("GET", "/api/v1/orgs/"+orgID+"/invitations", nil)
	if err != nil {
		return nil, err
	}
	return decode[[]Invitation](resp)
}
